docs/proposals/templates.md

```diff
@@ -0,0 +1,400 @@
+
+*The primary goal is that parameterization should facilitate reuse of declarative configuration templates in different environments in 
+  a "significant number" of common cases without further expansion, substitution, or other static preprocessing.*
+
+* This solution provides for templates that can be reused as is (assuming parameters are not used or provide sane default values) across 
+  different environments, they are a self-contained description of a topology.
+
+*Parameterization should not impede the ability to use kubectl commands with concrete resource specifications.*
+
+* The parameterization proposal here does not extend beyond Template objects.  That is both a strength and limitation of this proposal.
+  Parameterizable objects must be wrapped into a Template object, rather than existing on their own.
+
+*Parameterization should work with all kubectl commands that accept --filename, and should work on templates comprised of multiple resources.*
+
+* Same as above.
+
+*The parameterization mechanism should not prevent the ability to wrap kubectl with workflow/orchestration tools, such as Deployment manager.*
+
+* Since this proposal uses standard API objects, a DM or Helm flow could still be constructed around a set of templates, just as those flows are
+  constructed around other API objects today.
+
+*Any parameterization mechanism we add should not preclude the use of a different parameterization mechanism, it should be possible 
+to use different mechanisms for different resources, and, ideally, the transformation should be composable with other 
+substitution/decoration passes.*
+
+* This templating scheme does not preclude layering an additional templating mechanism over top of it.  For example, it would be 
+  possible to write a Mustache template which, after Mustache processing, resulted in a Template which could then be instantiated 
+  through the normal template instantiating process.
+
+*Parameterization should not compromise reproducibility. For instance, it should be possible to manage template arguments as well as 
+templates under version control.*
+
+* Templates are a single file, including default or chosen values for parameters.  They can easily me managed under version control.
+
+*It should be possible to specify template arguments (i.e., parameter values) declaratively, in a way that is "self-describing" 
+(i.e., naming the parameters and the template to which they correspond). It should be possible to write generic commands to 
+process templates.*
+
+* Parameter definitions include metadata which describes the purpose of the parameter.  Since parameter definitions are part of the template,
+  there is no need to indicate which template they correspond to.
+
+*It should be possible to validate templates and template parameters, both values and the schema.*
+
+* Template objects are subject to standard api validation.
```

Comment:
We should be able to validate the template field names.

As for values, if we had jsonschema-like validation specifications for parameters and/or for the resource fields, we could validate the values to some degree.